ethToElrond/v2/bridge/bridgeExecutor.go

```diff
@@ -0,0 +1,443 @@
+	executor.topologyProviderOnElrond = args.TopologyProviderOnElrond
+	executor.topologyProviderOnEthereum = args.TopologyProviderOnEthereum
+	executor.timeForTransferExecution = args.TimeForTransferExecution
+	return executor, nil
+}
+
+func checkBaseArgs(args ArgsBaseBridgeExecutor) error {
+	if check.IfNil(args.Log) {
+		return v2.ErrNilLogger
+	}
+	if check.IfNil(args.ElrondClient) {
+		return v2.ErrNilElrondClient
+	}
+	if check.IfNil(args.EthereumClient) {
+		return v2.ErrNilEthereumClient
+	}
+	return nil
+}
+
+func createBaseBridgeExecutor(args ArgsBaseBridgeExecutor) *bridgeExecutor {
+	return &bridgeExecutor{
+		log:            args.Log,
+		elrondClient:   args.ElrondClient,
+		ethereumClient: args.EthereumClient,
+	}
+}
+
+// GetLogger returns the logger implementation
+func (executor *bridgeExecutor) GetLogger() logger.Logger {
+	return executor.log
+}
+
+// MyTurnAsLeaderOnElrond returns true if the current relayer node is the leader on Elrond
+func (executor *bridgeExecutor) MyTurnAsLeaderOnElrond() bool {
+	return executor.topologyProviderOnElrond.MyTurnAsLeader()
+}
+
+// GetBatchFromElrond fetches the pending batch from Elrond
+func (executor *bridgeExecutor) GetBatchFromElrond(ctx context.Context) (*clients.TransferBatch, error) {
+	return executor.elrondClient.GetPending(ctx)
+}
+
+// StoreBatchFromElrond saves the pending batch from Elrond
+func (executor *bridgeExecutor) StoreBatchFromElrond(batch *clients.TransferBatch) error {
+	if batch == nil {
+		return v2.ErrNilBatch
+	}
+
+	executor.batch = batch
+	return nil
+}
+
+// GetStoredBatch returns the stored batch
+func (executor *bridgeExecutor) GetStoredBatch() *clients.TransferBatch {
+	return executor.batch
+}
+
+// GetLastExecutedEthBatchIDFromElrond returns the last executed batch ID that is stored on the Elrond SC
+func (executor *bridgeExecutor) GetLastExecutedEthBatchIDFromElrond(ctx context.Context) (uint64, error) {
+	return executor.elrondClient.GetLastExecutedEthBatchID(ctx)
+}
+
+// VerifyLastDepositNonceExecutedOnEthereumBatch will check the deposit nonces from the fetched batch from Ethereum client
+func (executor *bridgeExecutor) VerifyLastDepositNonceExecutedOnEthereumBatch(ctx context.Context) error {
+	if executor.batch == nil {
+		return v2.ErrNilBatch
+	}
+
+	lastNonce, err := executor.elrondClient.GetLastExecutedEthTxID(ctx)
+	if err != nil {
+		return err
+	}
+
+	return executor.verifyDepositNonces(lastNonce)
+}
+
+func (executor *bridgeExecutor) verifyDepositNonces(lastNonce uint64) error {
+	startNonce := lastNonce + 1
+	for _, dt := range executor.batch.Deposits {
+		if dt.Nonce != startNonce {
+			return fmt.Errorf("%w for deposit %s, expected: %d", v2.ErrInvalidDepositNonce, dt.String(), startNonce)
+		}
+
+		startNonce++
+	}
+
+	return nil
+}
+
+// GetAndStoreActionIDForProposeTransferOnElrond fetches the action ID for ProposeTransfer by using the stored batch. Returns and stores the action ID
```

Comment:
from a logical point of view, this stores first and then returns the action ID :smile: